lib/tree.rb

```diff
@@ -16,40 +16,127 @@ def initialize
+    #       if !current.left.nil?
+    #         current = current.left
+    #       else
+    #         current.left = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     else
+    #       if !current.right.nil?
+    #         current = current.right
+    #       else
+    #         current.right = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     end
+    #   end
+    # end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+
+    if key < current_node.key
+      find_helper(current_node.left, key)
+    else
+      find_helper(current_node.right, key)
+    end
+  end
+
+  # Time Complexity: O(logn) where n is the number of nodes
+  # Space Complexity: O(logn) where n is the number of nodes
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    return list
+  end
   # Time Complexity: 
   # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    # left -> current node -> right
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    # root -> left -> right
+    return list if current_node.nil?
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(logn) where n is the number of nodes
+  # Space Complexity: O(logn) where n is the number of nodes
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value }
+    return list
+  end
+
+  # Time Complexity: O(logn) where n is the number of nodes
+  # Space Complexity: O(logn) where n is the number of nodes
   def postorder
-    raise NotImplementedError
+    # left -> right -> root
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def height_helper(current_node, max, count)
+    return max if current_node.nil?
+
+    if count > max
+      max = count
+    end
+
+    height_helper(current_node.left, max, count + 1)
+    height_helper(current_node.right, max, count + 1)
+  end
+
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) where n is the number of nodes
   def height
```

Comment:
You are correct on time complexity, but the space complexity is O(log n) because you only make a max of O(log n) recursive calls, if the tree is balanced.

If the tree is imbalanced, you are right on space complexity.